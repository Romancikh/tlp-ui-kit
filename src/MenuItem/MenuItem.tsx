import MuiMenuItem from "@mui/material/MenuItem";
import { PropsWithChildren } from "react";

type MenuItemProps = PropsWithChildren & {
  onClick?: () => void;
  value?: string;
};

function MenuItem({ children, onClick, value }: MenuItemProps) {
  return (
    <MuiMenuItem onClick={onClick} value={value}>
      {children}
    </MuiMenuItem>
  );
}

export default MenuItem;
