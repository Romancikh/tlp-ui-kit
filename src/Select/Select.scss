.select {
  &__value {
    font-family: "Roboto", "Helvetica", "Arial", sans-serif;
    font-weight: 400;
    font-size: 1rem;
    line-height: 1.4375em;
    letter-spacing: 0.00938em;
    color: #000000de;
    box-sizing: border-box;
    display: inline-flex;
    align-items: center;
    position: relative;
    border-radius: 4px;
    border: 1px solid #0000003b;
    padding: 15px 8px 15px 13px;
    user-select: none;
    cursor: pointer;

    &:hover {
      border-color: #000000de;
    }

    &_active {
      border: 2px solid #1976d2;

      &:hover {
        border: 2px solid #1976d2;
      }
    }

    &_size_small {
      padding: 7px 7px 7px 13px;
    }

    &_full-width {
      width: 100%;
      justify-content: space-between;
    }

    &_disabled {
      color: #00000061;
      cursor: default;

      &:hover {
        border-color: #0000003b;
      }
    }
  }

  &__items {
    position: fixed;
    min-width: 108px;
  }
}

.icon {
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentcolor;
  flex-shrink: 0;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  right: 7px;
  top: calc(50% - 0.5em);
  pointer-events: none;
  color: rgba(0, 0, 0, 0.54);

  &_rotate {
    transform: rotate(180deg);
  }
}
