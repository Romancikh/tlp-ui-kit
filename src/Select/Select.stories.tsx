import type { Meta } from "@storybook/react";

import { useState } from "react";

import MenuItem from "../MenuItem/MenuItem";
import Select from "./Select";

const meta = {
  component: Select,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  title: "Select",
} satisfies Meta<typeof Select>;

export default meta;

export function Default() {
  const [value, setValue] = useState("Option 1");
  return (
    <Select onChange={setValue} value={value}>
      <MenuItem value="Option 1">Option 1</MenuItem>
      <MenuItem value="Option 2">Option 2</MenuItem>
      <MenuItem value="Option 3">Option 3</MenuItem>
      <MenuItem value="Option 4">Option 4</MenuItem>
    </Select>
  );
}
