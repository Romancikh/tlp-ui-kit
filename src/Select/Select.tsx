import {
  Children,
  PropsWithChildren,
  cloneElement,
  useRef,
  useState,
} from "react";
import Icon, { IconName } from "../Icon/Icon";
import List from "../List/List";
import Paper from "../Paper/Paper";
import "./Select.scss";

export type SelectChangeEvent<Value = string> =
  | (Event & { target: { value: Value; name: string } })
  | React.ChangeEvent<HTMLInputElement>;

export type SelectProps = PropsWithChildren & {
  size?: "small";
  fullWidth?: boolean;
  value?: string;
  disabled?: boolean;
  onChange?: (value: string) => void;
  maxWidth?: number;
  flex?: string;
};

function Select({
  size,
  fullWidth,
  value,
  disabled,
  onChange,
  maxWidth,
  flex,
  children,
}: SelectProps) {
  const [open, setOpen] = useState(false);
  const [iconName, setIconName] = useState<IconName>("ExpandMoreIcon");
  const selectValueRef = useRef<HTMLDivElement | null>(null);

  const handleSelect = (selectedValue: string) => {
    if (!onChange) return;
    onChange(selectedValue);
    setOpen(false);
    setIconName("ExpandMoreIcon");
  };

  return (
    <div className={`select`} style={{ flex }}>
      <div
        ref={selectValueRef}
        className={`select__value select__value_size_${size} select__value_${
          fullWidth ? "full-width" : ""
        } select__value_${disabled ? "disabled" : ""}`}
        style={{
          maxWidth,
        }}
        onClick={() => {
          if (disabled) return;
          setOpen(!open);
          if (iconName === "ExpandMoreIcon") setIconName("ExpandLessIcon");
          else setIconName("ExpandMoreIcon");
        }}
      >
        {value} <Icon name={iconName} />
      </div>
      {open && (
        <Paper
          elevation={10}
          className="select__items"
          width={selectValueRef.current?.offsetWidth}
        >
          <List>
            {Children.map(children, (child) =>
              cloneElement(child as React.ReactElement, {
                onClick: () => {
                  if (!child) return;
                  if (!(typeof child === "object" && "props" in child)) return;
                  handleSelect(child.props.value);
                },
              })
            )}
          </List>
        </Paper>
      )}
    </div>
  );
}

export default Select;
