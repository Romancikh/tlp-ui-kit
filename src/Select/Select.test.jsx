import { render } from "@testing-library/react";
import { useState } from "react";
import MenuItem from "../MenuItem/MenuItem";
import Select from "./Select";

const Component = () => {
  const [value, setValue] = useState("Option 1");

  return (
    <Select value={value} onChange={(event) => setValue(event.target.value)}>
      <MenuItem value="Option 1">Option 1</MenuItem>
      <MenuItem value="Option 2">Option 2</MenuItem>
      <MenuItem value="Option 3">Option 3</MenuItem>
      <MenuItem value="Option 4">Option 4</MenuItem>
    </Select>
  );
};

test("Select renders correctly", () => {
  const { container } = render(<Component />);
  expect(container).toMatchSnapshot();
});
