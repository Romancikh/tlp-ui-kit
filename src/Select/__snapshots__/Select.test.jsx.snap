// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders correctly 1`] = `
<div>
  <div
    class="select"
  >
    <div
      class="select__value select__value_size_undefined select__value_ select__value_ select__value_"
    >
      Option 1
       
      <svg
        aria-hidden="true"
        class="icon icon_"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
